File: solutions/85_maximal_rectangle.cpp
#include "aux.cpp"
#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <set>
#include <stdio.h>
#include <unordered_map>
#include <vector>
/**
   Given a 2D binary matrix filled with 0's and 1's, find the largest rectangle
   containing only 1's and return its area.
   Example:
   Input:
   [
   ["1","0","1","0","0"],
   ["1","0","1","1","1"],
   ["1","1","1","1","1"],
   ["1","0","0","1","0"]
   ]
   Output: 6

   https://leetcode.com/problems/maximal-rectangle/description/
**/
using namespace std;

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
class Solution {
public:
  int maximalRectangle(vector<vector<char>> &matrix) {
    if (0 == matrix.size())
      return 0;
    int res = 0;
    vector<vector<pair<int, int>>> cache(
        matrix.size(), vector<pair<int, int>>(matrix[0].size()));
    for (int i = matrix.size() - 1; i >= 0; i--) {
      for (int j = matrix[0].size() - 1; j >= 0; j--) {
        if (matrix[i][j] == '0')
          cache[i][j] = make_pair(0, 0);
        else if (i == matrix.size() - 1) {
          cache[i][j] = j == matrix[0].size() - 1
                            ? make_pair(1, 1)
                            : make_pair(cache[i][j + 1].first + 1, 1);
          res = max(res, cache[i][j].first);
        } else if (j == matrix[0].size() - 1) {
          cache[i][j] = i == matrix.size() - 1
                            ? make_pair(1, 1)
                            : make_pair(1, cache[i + 1][j].second + 1);
          res = max(res, cache[i][j].second);
        } else {
          cache[i][j] =
              make_pair(cache[i][j + 1].first + 1, cache[i + 1][j].second + 1);
          // cout << i << " " << j << " / ";
          // say(cache[i][j]);
          int rowlen = cache[i][j].first;
          for (int k = 0; k < cache[i][j].second; k++) {
            rowlen = min(rowlen, cache[i + k][j].first);
            res = max(res, (k + 1) * rowlen);
          }
        }
      }
    }
    return res;
  }
};

int main() {
  Solution s;
  vector<std::vector<char>> matrix = {{'1', '0', '1', '1', '1'},
                                      {'1', '0', '1', '1', '1'},
                                      {'1', '1', '1', '1', '1'},
                                      {'1', '1', '1', '1', '1'}};

  // matrix = {{'1', '1', '1', '1', '1', '1', '1', '1'},
  //           {'1', '1', '1', '1', '1', '1', '1', '0'},
  //           {'1', '1', '1', '1', '1', '1', '1', '0'},
  //           {'1', '1', '1', '1', '1', '0', '0', '0'},
  //           {'0', '1', '1', '1', '1', '0', '0', '0'}};

  say(s.maximalRectangle(matrix));
}

File: _site/solutions/85_maximal_rectangle.rb
# Given a 2D binary matrix filled with 0's and 1's, find the largest rectangle containing only 1's and return its area.
# Example:
# Input:
# [
#   ["1","0","1","0","0"],
#   ["1","0","1","1","1"],
#   ["1","1","1","1","1"],
#   ["1","0","0","1","0"]
# ]
# Output: 6
#
#  https://leetcode.com/problems/maximal-rectangle/description/
require './aux.rb'

# @param {Character[][]} matrix
# @return {Integer}
def maximal_rectangle(matrix)
  return 0 if matrix.empty?
  cache = Array.new(matrix.size) { Array.new(matrix.size, []) }
  res = 0
  (matrix.size - 1).downto(0).each do |i|
    (matrix[0].size - 1).downto(0).each do |j|
      if matrix[i][j] == '0'
        cache[i][j] = [0, 0]
      elsif i == matrix.size - 1
        cache[i][j] = j == matrix[0].size - 1 ? [1, 1] : [1 + cache[i][j + 1].first, 1]
        res = [res, cache[i][j].first].max
      elsif j == matrix[0].size - 1
        cache[i][j] = i == matrix.size - 1 ? [1, 1] : [1, 1 + cache[i + 1][j].last]
        res = [res, cache[i][j].last].max
      else
        cache[i][j] = [cache[i][j + 1].first + 1, cache[i + 1][j].last + 1]
        rowlen = cache[i][j].first
        i.upto(i + cache[i][j].last - 1).each do |ni|
          rowlen = [rowlen, cache[ni][j].first].min
          res = [(ni - i + 1) * rowlen, res].max
        end
      end
    end
  end
  res
end

matrix = [
  %w[1 0 1 0 0],
  %w[1 0 1 1 1],
  %w[1 1 1 1 1],
  %w[1 0 1 1 0]
]

p maximal_rectangle(matrix)
